marlowe-playground-client/src/Wallet.purs

```diff
@@ -0,0 +1,1301 @@
+        [ text $ "The contract is supposed to make a payment of "
+            <> show expected
+            <> " units of "
+            <> show tok
+            <> " from account "
+            <> show accNum
+            <> " of "
+            <> show owner
+            <> " to "
+            <> ( case payee of
+                  (Account (AccountId accNum2 owner2)) -> ("account " <> show accNum2 <> " of " <> show owner2)
+                  (Party dest) -> ("party " <> show dest)
+              )
+            <> " but there is only "
+            <> show amount
+            <> "."
+        ]
+    ]
+
+  displayWarning' (TransactionShadowing valId oldVal newVal) =
+    [ div [ classes [ rTableCell, first ] ] []
+    , div [ class_ (ClassName "RTable-2-cells") ] [ text "TransactionShadowing" ]
+    , div [ class_ (ClassName "RTable-4-cells") ]
+        [ text $ "The contract defined the value with id "
+            <> show valId
+            <> " before, it was assigned the value "
+            <> show oldVal
+            <> " and now it is being assigned the value "
+            <> show newVal
+            <> "."
+        ]
+    ]
+
+  displayWarning' TransactionAssertionFailed =
+    [ b_ [ text "TransactionAssertionFailed" ]
+    , text " - An assertion in the contract did not hold."
+    ]
+
+renderAssets :: forall p. State -> HTML p Action
+renderAssets state =
+  div []
+    ( [ h6 [ class_ noMargins ]
+          [ small [ classes [ textSecondaryColor, bold, uppercase ] ] [ text "Assets" ]
+          ]
+      ]
+        <> map currencyRow allTokens
+    )
+  where
+  emptyTokens = Map.fromFoldable $ Set.map (\t -> Tuple t zero) $ state ^. _tokens
+
+  walletTokens = state ^. (_openWallet <<< _Just <<< _assets)
+
+  allTokens = Map.toUnfoldable $ Map.union walletTokens emptyTokens
+
+  currencyRow :: Tuple Token BigInteger -> HTML p Action
+  currencyRow (Tuple token@(Token currencySymbol tokenName) money) =
+    div [ class_ (ClassName "currency-row") ]
+      ( [ div []
+            ( case currencySymbol, tokenName of
+                "", "" -> [ b_ [ text $ show money <> " Ada" ] ]
+                _, _ ->
+                  [ b_ [ text $ show money ]
+                  , spanText " "
+                  , b_ [ text currencySymbol ]
+                  , spanText " "
+                  , b_ [ text tokenName ]
+                  ]
+            )
+        ]
+          <> [ div []
+                [ addCurrencyInput
+                , addCurrencyButton
+                ]
+            ]
+      )
+    where
+    addCurrencyInput =
+      input
+        [ type_ InputNumber
+        , placeholder "BigInteger"
+        , class_ $ ClassName "currency-input"
+        , value "0"
+        , onValueChange
+            $ ( \newMoney ->
+                  Just
+                    $ ChangeCurrencyInput token
+                        ( fromMaybe zero (BigInteger.fromString newMoney)
+                        )
+              )
```

Comment:
Would this work as:

```purescript
\newMoney -> ChangeCurrentyInput token <$> BigInteger.fromString newMoney
```

I mean, the `Just` with the `fromMaybe` is suspicious.

Reply:
done.